Add capped results messaging & action/flow counts to lists

nmajor25 opened the pull request:
Shortcut Story ID: [sc-34741]

The backend recently capped list results to a maximum of 500 flows/actons. This PR is for adding messaging to lists on the frontend that indicates when a user's results are being capped.

The list count messaging can be in 4 different states:

1. The list count is below the capped maximum. ([screenshot](https://user-images.githubusercontent.com/35355575/230633166-9bc679b5-4dd3-4a38-bb72-aaf0149cdfff.png))
2. The capped maximum has been reached. ([screenshot](https://user-images.githubusercontent.com/35355575/230633296-faa7687b-80d0-4087-9e29-9a48074ef567.png))
3. The capped maximum has been reached, and the user narrows the list via the `Find in list...` input. ([screenshot](https://user-images.githubusercontent.com/35355575/230633351-df6688ad-a4eb-43df-a1b5-3337a25e712c.png))
4. The action/flow list is empty, so no count messaging is shown. ([screenshot](https://user-images.githubusercontent.com/35355575/230633239-20c180bb-1710-4c04-add8-b574856bc56d.png))

These changes apply to the worklists, schedule, and reduced schedule.

nmajor25 committed:
Add capped results messaging & list counts to worklists, schedule, & reduced schedule

The backend has caped results at 500 on lists, so the frontend needs messaging that tells the user when that cap is being applied

nmajor25 commented on src/js/views/patients/worklist/worklist_views.js:
```diff
@@ -222,6 +223,43 @@ const EmptyFindInListView = View.extend({
+    return {
+      maximumCount: this.getOption('maximumCount'),
+      hasReachedMaximum: this.getOption('hasReachedMaximum'),
+      count: this.getOption('count'),
```
Do we want to add thousand separator formatting to the `count` and `maximumCount` (i.e. `1000` => `1,000`? 

It's not a problem now. But in the future, the capped results could be more than 1,000?

Could just throw a `.toLocaleString()` on it.

Just a thought as I was reviewing this.

paulfalgout replied:
I don't think we need to worry about the separator for now.

nmajor25 commented on src/js/views/patients/worklist/worklist_views.js:
```diff
@@ -222,6 +223,43 @@ const EmptyFindInListView = View.extend({
+          {{formatMessage (intlGet "patients.worklist.worklistViews.countView.actionsCount") itemCount=count}}
+        {{/if}}
+      </span>
+    {{/if}}
```
This got a little complicated because the messaging needs to accommodate plural vs. non-plural wording (i.e. `"1 Action"` vs. `"1 Actions"`). Along with the different versions of `Flows` vs. `Actions`.

paulfalgout replied:
I think in the case of lots of `if` `else` templates, we can clean it up a bit by using that logic in a `getTemplate` and returning smaller simpler templates for each (or  maybe just some) logic branch.

nmajor25 replied:
Updated these to use a `getTemplate()`. Should be simpler now, but open to any other ways to simplify them more.

nmajor25 commented on src/js/static.js:
```diff
@@ -1,3 +1,5 @@
+const MAXIMUM_LIST_COUNT = /* istanbul ignore next */ _TEST_ ? 30 : 500;
```
Set to a maximum of `500` in production. And `30` in Cypress.

Loading a list of `500` actions/flows in Cypress was making me nervous that it would cause issues. 

The largest list I could find in the tests was `20`, so I figured `30` was safe. But maybe we want to raise it to `50` or something a little higher to be sure other tests don't unknowingly reach that max in the future... 🤔

paulfalgout replied:
not a bad idea.  if this number changes in the future, let's request it get put on the meta maybe?  I'm not 100% sure how we're going to get back that result or handle it, but for now I think your static option works well.

nmajor25 replied:
Sounds good, we'll wait for it to change in the future.

I also updated the maximum list count for Cypress from `30` => `50`. Should reduce the chance of accidentally hitting that limit in future tests.

nmajor25 commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -258,6 +266,25 @@ export default App.extend({
+      count: this.filteredCollection.length,
+      maximumCount: MAXIMUM_LIST_COUNT,
+      isFindInListApplied: this.getState('searchQuery').length > 2,
+      hasReachedMaximum: this.collection.length === MAXIMUM_LIST_COUNT,
```
The capped messaging will only show if the list count is exactly equal to the maximum count (i.e. `500`). Which is based on the `MAXIMUM_LIST_COUNT` static variable.

We'll have to keep the maximum list count in-sync with what the backend is using.

Would it be a good idea to have the actions/flows API requests include a `maxCount: 500` item or something similar?

Than we wouldn't have to update `MAXIMUM_LIST_COUNT` manually if the backend changes the maximum count on their end.

nmajor25 commented on src/js/apps/patients/schedule/reduced_schedule_app.js:
```diff
@@ -38,16 +40,23 @@ export default App.extend({
+    this.listenTo(collectionView, 'filtered', filtered => {
+      this.filteredCollection.reset(filtered);
+      this.showCountView();
+    });
```
In order for the count messaging to be updated with the `Find in list...`, I needed to add this `'filtered'` listener to the list and the `this.filteredCollection` that is used in the other schedule and worklist apps.

nmajor25 commented:
A few things I noticed when working on this that need to be explored further:

1. On the schedule, actions without a due date are removed from the list (via `groupByDate()` [here](https://github.com/RoundingWell/care-ops-frontend/blob/develop/src/js/apps/patients/schedule/schedule_app.js#L146)). But that happens after the API request is retrieved.
    - So there could be a scenario where the schedule retrieves results that are capped at `500`. But then the `groupByDate()` removes say `100` actions without a due date. So the schedule results would end up only showing `400` actions.
    - Can we update the API request to only retrieve actions with a due date that is not `null`? 
2. When the count messaging is shown/hidden on the reduced schedule, there's a page layout jump ([screengrab](https://user-images.githubusercontent.com/35355575/230641807-cc6c8e7f-c88e-4dca-af91-cec7c8508874.mov)). We could set a fixed height to that to avoid the page jump... or maybe there's a better way. But still need to look into that.

paulfalgout commented on src/js/apps/patients/schedule/schedule_app.js:
```diff
@@ -214,6 +222,24 @@ export default App.extend({
+      count: this.filteredCollection.length,
+      maximumCount: MAXIMUM_LIST_COUNT,
+      isFindInListApplied: this.getState('searchQuery').length > 2,
+      hasReachedMaximum: this.collection.length === MAXIMUM_LIST_COUNT,
```
these same calculations are in a few places.  I think this view could be passed just `this.collection` and `this.filteredCollection` 

Then the view can import `MAXIMUM_LIST_COUNT` and find in list can be implied if `this.filteredCollection.length < this.collection.length` which might be better anyhow as 500 of 500 isn't entirely useful.

nmajor25 replied:
That seems simpler. Updated to that in my last push.

paulfalgout commented on src/js/views/patients/schedule/schedule_views.js:
```diff
@@ -111,6 +112,33 @@ const SelectAllView = View.extend({
+      </div>
+      <div>{{ @intl.patients.worklist.worklistViews.countView.narrowFilters }}</div>
+    {{else}}
+      <span class="schedule__count-total">
```
I think we could probably use `<strong>` here

nmajor25 replied:
Yup, good call. Switched to that in the last push.

paulfalgout commented on test/integration/patients/worklist/schedule.js:
```diff
@@ -1709,6 +1929,12 @@ context('schedule page', function() {
+    cy
+      .get('[data-count-region] .schedule__count-total')
+      .should(el => {
+        expect(el.text().trim()).to.equal('1 Action');
```
why is the trim necessary?  `11 Actions`?
could maybe be `should('contain', '1 Action').should('not.contain', 'Actions')`

nmajor25 replied:
I'm not sure why the `trim()` is/was necessary, it had a bunch of spaces added when Cypress was comparing them.

But `should('contain', '1 Action').should('not.contain', 'Actions')` is a better way to do it. So I switched it over to that instead.

paulfalgout commented on src/js/views/patients/worklist/worklist_views.js:
```diff
@@ -222,6 +223,43 @@ const EmptyFindInListView = View.extend({
   `,
 });
 
+const CountView = View.extend({
```
could see a world where this view is in some `patient/shared` directory.  I know the flow doesn't apply everywhere, but if that flag isn't set, it isn't set..  otherwise it is all identical correct?

nmajor25 replied:
Ya, it should all be identical besides the flow wording. I pushed the other changes up, but I'm still working on this.

nmajor25 commented:
Reason for this commit: https://github.com/RoundingWell/care-ops-frontend/compare/ed8533d52a03a5886a9ef3199849087d42445934..12e03119648a547f1c6dd5f22c5b84c70bc474b2.

In the worklist test I added, I was using the `Action 50`/`Flow 50` and `Action 5`/`Flow 5` find in list queries to test the count messaging being shown if it's narrowed to show `Showing 1 Action of 50.` or `Showing 2 Actions of 50.`.

But this will cause flakiness if the tests run at the wrong time because of the `Added`/`Updated` list columns being included in the FIL filter. 

For example, the test would fail if the test ran at 9:50am, because each action would be included in the `Action 50` FIL query when it's only supposed to include one.

To fix this, I updated the test so the first action/flow has a name of `First Action`/`First Flow` (used to test `Showing 1 Action of 50.` messaging). And I made it so every other action/flow has `Test Patient`/`Other Patient` (used to test `Showing 25 Actions of 50.` messaging with the `Test Patient` FIL query).

This shouldn't be a problem on the schedule, but I updated those tests to match the worklist one.

paulfalgout commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -258,6 +264,23 @@ export default App.extend({
     Radio.request('alert', 'show:success', renderTemplate(BulkEditActionsSuccessTemplate, { itemCount }));
   },
+  showCountView() {
+    if (!this.collection || !this.filteredCollection.length) {
```
you know what else would work here?  Since the `CountView` is pretty simple if could do this `!this.collection || !this.filteredCollection.length` and then just render an empty template?  I think a `<div></div>` isn't going to hurt anything?

paulfalgout commented on src/js/views/patients/worklist/worklist_views.js:
```diff
@@ -222,6 +225,62 @@ const EmptyFindInListView = View.extend({
+`;
+
+const MaximumCountTemplate = hbs`
+  <div>{{formatMessage (intlGet "patients.worklist.worklistViews.countView.maximumListCount") maximumCount=maximumCount listTypeLabel=listTypeLabel}}</div>
```
these don't need the wrapping `<div>`s

nmajor25 replied:
The `<div>`s are what make the two lines stack.

Without `<div>`s:

<img width="274" alt="Screenshot 2023-04-11 at 9 49 36 AM" src="https://user-images.githubusercontent.com/35355575/231201209-f34b5256-67d7-43a9-97ea-fe3d4a4614a1.png">

Or just remove the `<div>` around the `{{ formatMessage ... }}` portion?

paulfalgout replied:
oh no you're right.. the line wrapping just made it look like the whole template was wrapped in a div.

paulfalgout reviewed (changes requested):
Yeah looks good.. FWIW it doesn't have to be a component to be shared.  We have also have `_views` files in some shared folders.

paulfalgout commented on src/js/views/patients/worklist/worklist_views.js:
```diff
@@ -222,6 +225,64 @@ const EmptyFindInListView = View.extend({
+      maximumCount: MAXIMUM_LIST_COUNT,
+      count: filteredCollection.length,
+      isFlowList,
+      listTypeLabel: isFlowList ? 'Flows' : 'Actions',
```
can't do this.  needs to be in the i18n.. but the i18n could have a `{listTypeLabel, select, true "Flows" false "Actions"}` or something like that?  I dunno if it works with boolean..

nmajor25 replied:
Good catch, just pushed an update for that.

paulfalgout commented on src/js/views/patients/schedule/schedule_views.js:
```diff
@@ -111,6 +114,51 @@ const SelectAllView = View.extend({
   },
 });
 
+const ListCountTemplate = hbs`
```
can we get away with not using any of these views/templates and moving the worklist ones to a shared location?
